### PR TITLE
[AIR] <Part 4> Lightning Trainer Release tests + docstring sample test

### PR DESCRIPTION
**Pull request by woshiyyya**
## Why are these changes needed?

This PR aims to add release tests and docstring tests for `LightningTrainer`.

The release tests:
- Cluster setting: 3 x g4dn.4xlarge (48 CPUs + 3 GPUs)
- **lightning_gpu_train_3x16_3x1**: 
  - Test end-to-end training on MNIST dataset
  -  DDP training with 3 GPUs 
- **lightning_gpu_tune_3x16_3x1**: 
  - Test LightningTrainer integration with Ray Tune
  - 4 Trails + PBT scheduler

Release Test Job succeed:
- [lightning_gpu_train_3x16_3x1](https://console.anyscale-staging.com/o/anyscale-internal/jobs/prodjob_ygn4lsdw7e2s96mmfvrzuwlha3)
- [lightning_gpu_tune_3x16_3x1](https://console.anyscale-staging.com/o/anyscale-internal/jobs/prodjob_fva2ant7wv8aurytrht44dj7ti)
## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: release/ml_user_tests/lightning/driver_setup.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+cd "${0%/*}" || exit 1
+
+# Install extras (e.g. gym[atari]) are not allowed in constraint files.
+# Hack to remove the first line and all brackets in `requirements_tune.txt`
+# so that requirements_tune.txt can be used as a constraint in driver_requirements.txt
+sed '1d' ./../../../python/requirements/ml/requirements_tune.txt | sed 's/\[//g;s/\]//g' > tune_constraints.txt
```
Comment (gjoliver) [inline release/ml_user_tests/lightning/driver_setup.sh:8]: :)
why do you need atari? can we use a simpler workload for this test?

File: python/ray/train/lightning/lightning_trainer.py [hunk 2]
```diff
@@ -178,10 +178,8 @@ class LightningTrainer(TorchTrainer):
     using the arguments provided in ``trainer_init_config`` and then run
     ``pytorch_lightning.Trainer.fit``.
 
-    TODO(yunxuanx): make this example testable
-
     Example:
-        .. code-block:: python
+        .. testcode::
```
Comment (gjoliver) [inline python/ray/train/lightning/lightning_trainer.py:182]: how does this work? can you explain this a bit for me? I am curious.

File: release/ml_user_tests/lightning/driver_setup.sh (hunk 1 above)
Comment (woshiyyya) [reply release/ml_user_tests/lightning/driver_setup.sh:8]: Ahh actually not, I copied from ray-lightning test configs. Let me remove these.

File: python/ray/train/lightning/lightning_trainer.py (hunk 2 above)
Comment (woshiyyya) [reply python/ray/train/lightning/lightning_trainer.py:182]: I am also new to this. I asked Justin and Matt before, and it seems that the logic is hidden quite deeply. As far as I know, the code block within this tag is extracted and tested using an `m5.2xlarge` instance. The easiest way to obtain the testing configuration is to go to the Buildkite page and check the Environment tab.

**Review by amogkam — COMMENTED** [review]
Thanks! Can we also remove the existing ray lightning release test if it is being subsumed by the new release test?

File: python/ray/train/lightning/lightning_trainer.py [hunk 3]
```diff
@@ -272,6 +270,13 @@ def configure_optimizers(self):
                 scaling_config=scaling_config,
             )
             results = trainer.fit()
+            print(results)
```
Comment (justinvyu) [inline python/ray/train/lightning/lightning_trainer.py:273]: Should just be a single `result`?

File: release/lightning_tests/compute_tpl.yaml [hunk 4]
```diff
@@ -0,0 +1,22 @@
+cloud_id: {{env["ANYSCALE_CLOUD_ID"]}}
+region: us-west-2
+
+max_workers: 2
```
Comment (justinvyu) [inline release/lightning_tests/compute_tpl.yaml:4]: Not sure how needed this is. I see it in a bunch of places but seems redundant.

File: release/lightning_tests/compute_tpl.yaml [hunk 5]
```diff
@@ -0,0 +1,22 @@
+cloud_id: {{env["ANYSCALE_CLOUD_ID"]}}
```
Comment (justinvyu) [inline release/lightning_tests/compute_tpl.yaml:1]: Add a small comment at the top to describe what the compute is for this test? (N x node_type = X cpus, Y gpus total)

File: release/release_tests.yaml [hunk 6]
```diff
@@ -1010,6 +1010,47 @@
...
+    cluster_env: app_config.yaml
+    cluster_compute: compute_tpl.yaml
+
+  run:
+    timeout: 1200
+    script: python workloads/test_trainer.py
+    wait_for_nodes:
+      num_nodes: 3
+
+  alert: default
```
Comment (justinvyu) [inline release/release_tests.yaml:1034]: What does this do? 😮  Should it be a special "lightning" label like some xgboost tests do?

File: release/release_tests.yaml [hunk 7]
```diff
@@ -1010,6 +1010,47 @@
...
 
+#######################
+# Lightning tests
+#######################
+
+- name: lightning_trainer_test_latest
+  group: Lightning tests
+  working_dir: lightning_tests
+
+  frequency: nightly-3x
```
Comment (justinvyu) [inline release/release_tests.yaml:1021]: Is `nightly-3x` 3x per week? And `nightly` is 5x per week?

File: release/release_tests.yaml [hunk 8]
```diff
@@ -1010,6 +1010,47 @@
 
   alert: default
 
+#######################
+# Lightning tests
+#######################
+
+- name: lightning_trainer_test_latest
```
Comment (justinvyu) [inline release/release_tests.yaml:1017]: Do we need the `latest` tag at the end? What is `latest`: Ray, PTL?

What about this naming pattern:

```
lightning_gpu_3x16+1  <-- 3 x (16 CPU + 1 GPU)
lightning_gpu_tune_3x16+1  
```

This way, you can filter out tests based on prefix/suffix:

```
name:lightning_gpu_tune.*
name:lightning_gpu.*
```

Also, can add `cpu` only option in the future.

Comment (justinvyu) [reply release/release_tests.yaml:1017]: And add a comment about the naming in the yaml

File: release/lightning_tests/workloads/test_tuner.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    tuner = tune.Tuner(
+        lightning_trainer,
+        param_space={"lightning_config": lightning_config},
+        run_config=ray.air.RunConfig(
+            name="release-tuner-test",
+            verbose=2,
+            checkpoint_config=CheckpointConfig(
+                num_to_keep=1,
+                checkpoint_score_attribute="ptl/val_accuracy",
+                checkpoint_score_order="max",
```
Comment (justinvyu) [inline release/lightning_tests/workloads/test_tuner.py:129]: Isn't this a bit dangerous 😅

File: release/lightning_tests/workloads/test_trainer.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    trainer = LightningTrainer(
+        lightning_config=lightning_config,
+        scaling_config=scaling_config,
+    )
+
+    trainer.fit()
+
+    taken = time.time() - start
+    result = {
+        "time_taken": taken,
```
Comment (justinvyu) [inline release/lightning_tests/workloads/test_trainer.py:130]: Can we report some metrics about the training as part of job output? Like accuracy.

File: release/lightning_tests/workloads/test_tuner.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+            ),
+        ),
+        tune_config=tune.TuneConfig(
+            metric="ptl/val_accuracy", mode="max", num_samples=2
+        ),
+    )
+    results = tuner.fit()
+    print(results.get_best_result(metric="ptl/val_accuracy", mode="max"))
+    print(results.__dict__)
```
Comment (justinvyu) [inline release/lightning_tests/workloads/test_tuner.py:138]:
```suggestion
```

**Review by justinvyu — CHANGES_REQUESTED** [review]
Looks good! Have a few suggestions.

A few other things:
- We can combine `test_tuner` and `test_trainer`. Just switch between the two with a command line arg.
- For `test_tuner`, can we also use a scheduler like PBT? I know we don't want to add too much to one test, but this way we'd test out the PTL checkpointing/restoration logic in the release test.

File: release/lightning_tests/compute_tpl.yaml (hunk 5 above)
Comment (woshiyyya) [reply release/lightning_tests/compute_tpl.yaml:1]: sounds good

File: release/lightning_tests/workloads/test_tuner.py (hunk 9 above)
Comment (woshiyyya) [reply release/lightning_tests/workloads/test_tuner.py:129]: Oh wait, have we resolved this num_to_keep=1 problem?

File: release/release_tests.yaml (hunk 6 above)
Comment (woshiyyya) [reply release/release_tests.yaml:1034]: hmm, good question. 80% release tests use default. Let me check out the meanings of these keywords.

**Review by gjoliver — COMMENTED** [review]
looks solid. can you respond to all of Justin's comments?
also do you know, or maybe you can ask Kai quickly, if we can release test a script in our doc/examples folder?
that way, we don't have to duplicate code.

File: release/lightning_tests/workloads/test_tuner.py (hunk 9 above)
Comment (justinvyu) [reply release/lightning_tests/workloads/test_tuner.py:129]: I don't think so

**Comment by justinvyu** [comment]
@gjoliver I think it has to be in `release` folder. But, you can create a relative symlink to in the doc/examples folder.

**Review by justinvyu — APPROVED** [review]

**Review by gjoliver — APPROVED** [review]
